### PR TITLE
S5753 subtask - add IT with web.config in DirectoryBuild.props outside of the solution folder

### LINKED ISSUE #4112: S5753 subtask - add IT with web.config in DirectoryBuild.props outside of the solution folder
*Important*: before anything,  should also re-enable the existing IT `WebConfigTest#should_raise_hotspot_on_web_config()` (see #4125).

This issue should add an additional integration test with the web.config referenced in the DirectoryBuild.props in a parent folder.

The structure could be:
- folder/DirectoryBuild.props
- folder/Project1/Project1.sln
- folder/Project2/Project2.sln

And the analysis should be done inside `folder/Project1`. 

Like the other Java ITs, this will only work after the Scanner for .NET 5.1 gets released.

This issue is a sub-task for #3951 and should not be part of the milestone and should be merged to PR#3938 (on the feature branch)

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #4112

Java ITs will fail for now, but it can be reviewed.
Rebase will be needed when the rule implementation is updated in #4111.

I manually verified that `.sonarqube/conf/0` directory contains expected `SonarProjectConfig.xml`, new location for `FilesToAnalyze.txt` and `FilesToAnalyze.txt` contains `PathOutsideProjectRoot`.

Having `Directory.Build.props` fills our needs, as the build is run for a project (`WebConfig.*.csproj`/`WebConfig.*.vbproj`) directory. And we need a common root (SLN dir) for the files to be included in the SQ.

### CODE REVIEW

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Overall it looks good. What I would suggest is to add a web.config file in a subfolder of the project  and not include it in the build process. That way we will verify that only files included in the build process are collected and analyzed.
P.S. maybe even add some  weird casing files like wEb.cOnfIg?

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM

**Comment by pavel-mikula-sonarsource** [comment]
@csaba-sagi-sonarsource Only last 4 commits news a review
* Fix ITs project layout
* Fix AD001 when directory does not exist
* Swap expected
* Swap slashes
Java ITs are still failing duo to unrelated issue - we're not clearing cache property. Andrei will create another PR to fix that.

One more rebase with no changes will be needed so it can be reviewed already.

**Comment by pavel-mikula-sonarsource** [comment]
Latest commits fixes these comments for the feature branch as well:
https://github.com/SonarSource/sonar-dotnet/pull/3938#discussion_r592568962
https://github.com/SonarSource/sonar-dotnet/pull/3938#discussion_r593000291

I also needed a new commits, as rebase didn't re-trigger the CI

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DisablingRequestValidationBase.cs [hunk 1]
```diff
@@ -83,7 +83,7 @@ protected override void Initialize(SonarAnalysisContext context)
                     return;
                 }
 
-                foreach (var fullPath in context.ProjectConfiguration(c.Options).FilesToAnalyze.FindFiles("web.config"))
+                foreach (var fullPath in context.ProjectConfiguration(c.Options).FilesToAnalyze.FindFiles("web.config").Where(File.Exists))
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DisablingRequestValidationBase.cs:86]: How is it possible that we get some files from the scanner that do not exist?

File: its/src/test/java/com/sonar/it/vbnet/WebConfigTest.java [hunk 2]
```diff
@@ -39,17 +39,14 @@ public void init() {
   }
 
   @Test
-  @Ignore("Ignore until https://github.com/SonarSource/sonar-dotnet/issues/4111 gets fixed")
   public void should_raise_hotspot_on_web_config() throws Exception {
     final String projectName = "WebConfig.Vb";
```
Comment (andrei-epure-sonarsource) [inline its/src/test/java/com/sonar/it/vbnet/WebConfigTest.java:43]: @pavel-mikula-sonarsource should this be updated to `WebConfig.VB`?

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DisablingRequestValidationBase.cs (hunk 1 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DisablingRequestValidationBase.cs:86]: it's a good question, and it's also good to be defensive.
But it's a good question that we need to ask.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DisablingRequestValidationBase.cs:86]: If you mess up your project file or targets file, you can explicitly include a file that's not on the disk, but will be in the MsBuild list of files. From there, it gets propagated to FilesToAnalyze and we have fun with those.

File: its/src/test/java/com/sonar/it/vbnet/WebConfigTest.java (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply its/src/test/java/com/sonar/it/vbnet/WebConfigTest.java:43]: Yes, it can be aligned. Will do

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM once more.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

**Comment by pavel-mikula-sonarsource** [comment]
And this time, I mean it 😆

File: its/src/test/java/com/sonar/it/shared/TestUtils.java [hunk 3]
```diff
@@ -270,11 +267,14 @@ public static TemporaryFolder createTempFolder() {
     return folder;
   }
 
-  private static void deleteLocalCache() {
+  public static void deleteLocalCache() {
     // SonarScanner for .NET caches the analyzer, so running the test twice in a row means the old binary is used.
-    String localAppData = System.getenv("LOCALAPPDATA") + "\\Temp\\.sonarqube";
+    File file = new File(System.getenv("LOCALAPPDATA") + "\\Temp\\.sonarqube");
```
Comment (pavel-mikula-sonarsource) [inline its/src/test/java/com/sonar/it/shared/TestUtils.java:272]: I've verified that this path is enough for CI as well.

File: its/src/test/java/com/sonar/it/csharp/UnitTestProjectTypeProbingTest.java [hunk 4]
```diff
@@ -73,8 +73,10 @@ public void testProject_WithPropertySetToFalse_IsIdentifiedAsMainProjectByScanne
   }
 
   @Test
-  public void msTestProject_IsIdentifiedAsTestProject() {
-    assertThat(getIssues("UTProjectProbing:UTProjectProbing.MsTest/calculator.cs")).isEmpty();
+  public void msTestProject_IsIdentifiedAsMainProjectBzScannerAndTestByAnalyzer() {
```
Comment (andrei-epure-sonarsource) [inline its/src/test/java/com/sonar/it/csharp/UnitTestProjectTypeProbingTest.java:76]: nitpick
msTestProject_IsIdentifiedAsMainProjectByScannerAndTestByAnalyzer
(not BzScanner)

**Comment by pavel-mikula-sonarsource** [comment]
This PR is coursed 

>Downloaded 947.0 MB out of 1,207.5 MB (78%).
>Downloaded 1,184.1 MB out of 1,207.5 MB (98%).
>ApplicationInsightsTelemetrySender correlated 3 events with X-TFS-Session 7b6c3f4b-8d7c-43a3-a274-de8052c7d0e7
>##[error]The pipe has been ended.

**Pull request closed without merging**